File: server/kotlin/src/main/kotlin/com/lebonpoint/infrastructure/http/models/ItemDTO.kt
package com.lebonpoint.infrastructure.http.models

import com.lebonpoint.domain.entities.CreateItemData
import com.lebonpoint.domain.entities.Item
import com.lebonpoint.domain.entities.PatchItemData
import com.lebonpoint.domain.entities.ReplaceItemData
import com.lebonpoint.domain.valueobjects.ItemCondition
import com.lebonpoint.domain.valueobjects.ItemImage
import com.lebonpoint.domain.valueobjects.ItemStatus
import com.lebonpoint.shared.serializers.InstantSerializer
import kotlinx.serialization.Serializable
import java.time.Instant

/**
 * Health check response
 */
@Serializable
data class HealthResponse(
    var status: String,
    @Serializable(with = InstantSerializer::class)
    var timestamp: Instant = Instant.now()
)

/**
 * Item response for API
 */
@Serializable
data class ItemResponse(
    var id: Int,
    var title: String,
    var description: String? = null,
    var price_cents: Int,
    var category: String? = null,
    var condition: ItemCondition,
    var status: ItemStatus,
    var is_featured: Boolean,
    var city: String? = null,
    var postal_code: String? = null,
    var country: String,
    var delivery_available: Boolean,
    var created_at: String,
    var updated_at: String,
    var published_at: String? = null,
    var images: List<ItemImageResponse> = emptyList()
)

/**
 * Item image response
 */
@Serializable
data class ItemImageResponse(
    var url: String,
    var alt: String? = null,
    var sort_order: Int = 0
)

/**
 * List items response
 */
@Serializable
data class ListItemsResponse(
    var items: List<ItemResponse>,
    var next_cursor: String? = null
)

/**
 * Create item request
 */
@Serializable
data class CreateItemRequest(
    var title: String,
    var description: String? = null,
    var price_cents: Int,
    var category: String? = null,
    var condition: ItemCondition,
    var status: ItemStatus = ItemStatus.DRAFT,
    var is_featured: Boolean = false,
    var city: String? = null,
    var postal_code: String? = null,
    var country: String = "FR",
    var delivery_available: Boolean = false,
    var images: List<ItemImageRequest> = emptyList()
)

/**
 * Item image request
 */
@Serializable
data class ItemImageRequest(
    var url: String,
    var alt: String? = null,
    var sort_order: Int = 0
)

/**
 * Update item request (PUT - full replacement)
 */
@Serializable
data class UpdateItemRequest(
    var title: String,
    var description: String? = null,
    var price_cents: Int,
    var category: String? = null,
    var condition: ItemCondition,
    var status: ItemStatus,
    var is_featured: Boolean,
    var city: String? = null,
    var postal_code: String? = null,
    var country: String,
    var delivery_available: Boolean,
    var images: List<ItemImageRequest> = emptyList()
)

/**
 * Patch item request (PATCH - partial update)
 */
@Serializable
data class PatchItemRequest(
    var title: String? = null,
    var description: String? = null,
    var price_cents: Int? = null,
    var category: String? = null,
    var condition: ItemCondition? = null,
    var status: ItemStatus? = null,
    var is_featured: Boolean? = null,
    var city: String? = null,
    var postal_code: String? = null,
    var country: String? = null,
    var delivery_available: Boolean? = null,
    var images: List<ItemImageRequest>? = null
)

/**
 * Error response
 */
@Serializable
data class ErrorResponse(
    var error: ErrorDetail
)

/**
 * Error detail
 */
@Serializable
data class ErrorDetail(
    var code: String,
    var message: String,
    var details: Map<String, String?> = emptyMap()
)

/**
 * Convert Item entity to ItemResponse
 */
fun Item.toResponse(): ItemResponse {
    return ItemResponse(
        id = id,
        title = title,
        description = description,
        price_cents = priceCents,
        category = category,
        condition = condition,
        status = status,
        is_featured = isFeatured,
        city = city,
        postal_code = postalCode,
        country = country,
        delivery_available = deliveryAvailable,
        created_at = createdAt.toString(),
        updated_at = updatedAt.toString(),
        published_at = publishedAt?.toString(),
        images = images.map { it.toResponse() }
    )
}

/**
 * Convert ItemImage to ItemImageResponse
 */
fun ItemImage.toResponse(): ItemImageResponse {
    return ItemImageResponse(
        url = url,
        alt = alt,
        sort_order = sortOrder
    )
}

/**
 * Convert CreateItemRequest to CreateItemData
 */
fun CreateItemRequest.toData(): CreateItemData {
    return CreateItemData(
        title = title,
        description = description,
        priceCents = price_cents,
        category = category,
        condition = condition,
        status = status,
        isFeatured = is_featured,
        city = city,
        postalCode = postal_code,
        country = country,
        deliveryAvailable = delivery_available,
        images = images.map { it.toDomain() }
    )
}

/**
 * Convert UpdateItemRequest to ReplaceItemData
 */
fun UpdateItemRequest.toData(): ReplaceItemData {
    return ReplaceItemData(
        title = title,
        description = description,
        priceCents = price_cents,
        category = category,
        condition = condition,
        status = status,
        isFeatured = is_featured,
        city = city,
        postalCode = postal_code,
        country = country,
        deliveryAvailable = delivery_available,
        images = images.map { it.toDomain() }
    )
}

/**
 * Convert PatchItemRequest to PatchItemData
 */
fun PatchItemRequest.toData(): PatchItemData {
    return PatchItemData(
        title = title,
        description = description,
        priceCents = price_cents,
        category = category,
        condition = condition,
        status = status,
        isFeatured = is_featured,
        city = city,
        postalCode = postal_code,
        country = country,
        deliveryAvailable = delivery_available,
        images = images?.map { it.toDomain() }
    )
}

/**
 * Convert ItemImageRequest to ItemImage
 */
fun ItemImageRequest.toDomain(): ItemImage {
    return ItemImage(
        url = url,
        alt = alt,
        sortOrder = sort_order
    )
}

File: server/go/internal/application/dto/item_dto.go
package dto

import (
	"time"

	"github.com/hymaia/lebonpoint-app/server/go/internal/domain/valueobject"
)

// ItemResponse represents an item response
type ItemResponse struct {
	ID                int64                     `json:"id"`
	Title             string                    `json:"title"`
	Description       string                    `json:"description,omitempty"`
	PriceCents        int                       `json:"priceCents"`
	Category          string                    `json:"category,omitempty"`
	Condition         valueobject.ItemCondition `json:"condition"`
	Status            valueobject.ItemStatus    `json:"status"`
	IsFeatured        bool                      `json:"isFeatured"`
	City              string                    `json:"city,omitempty"`
	PostalCode        string                    `json:"postalCode,omitempty"`
	Country           string                    `json:"country"`
	DeliveryAvailable bool                      `json:"deliveryAvailable"`
	CreatedAt         time.Time                 `json:"createdAt"`
	UpdatedAt         time.Time                 `json:"updatedAt"`
	PublishedAt       *time.Time                `json:"publishedAt,omitempty"`
	Images            []valueobject.ItemImage   `json:"images"`
}

// FromEntity creates an ItemResponse from a domain entity
func (ir *ItemResponse) FromEntity(item interface{}) {
	// This would be implemented if we have a common entity interface
	// For now, we'll convert directly in handlers
}

// CreateItemRequest represents a request to create an item
type CreateItemRequest struct {
	Title             string                    `json:"title" binding:"required,min=3,max=200"`
	Description       string                    `json:"description,omitempty"`
	PriceCents        int                       `json:"priceCents" binding:"required,gte=0"`
	Category          string                    `json:"category,omitempty"`
	Condition         valueobject.ItemCondition `json:"condition" binding:"required"`
	Status            valueobject.ItemStatus    `json:"status" binding:"required,oneof=draft active reserved sold archived"`
	IsFeatured        bool                      `json:"isFeatured"`
	City              string                    `json:"city,omitempty"`
	PostalCode        string                    `json:"postalCode,omitempty"`
	Country           string                    `json:"country" binding:"required"`
	DeliveryAvailable bool                      `json:"deliveryAvailable"`
	Images            []valueobject.ItemImage   `json:"images"`
}

// GetDefaultValues returns default values for optional fields
func (req *CreateItemRequest) GetDefaultValues() map[string]interface{} {
	defaults := make(map[string]interface{})

	if req.Status == "" {
		defaults["status"] = valueobject.ItemStatusDraft
	}
	if req.Country == "" {
		defaults["country"] = "FR"
	}
	if req.Images == nil {
		defaults["images"] = []valueobject.ItemImage{}
	}

	return defaults
}

// UpdateItemRequest represents a request to update an item (full replacement)
type UpdateItemRequest struct {
	Title             string                    `json:"title" binding:"required,min=3,max=200"`
	Description       string                    `json:"description,omitempty"`
	PriceCents        int                       `json:"priceCents" binding:"required,gte=0"`
	Category          string                    `json:"category,omitempty"`
	Condition         valueobject.ItemCondition `json:"condition" binding:"required"`
	Status            valueobject.ItemStatus    `json:"status" binding:"required,oneof=draft active reserved sold archived"`
	IsFeatured        bool                      `json:"isFeatured"`
	City              string                    `json:"city,omitempty"`
	PostalCode        string                    `json:"postalCode,omitempty"`
	Country           string                    `json:"country" binding:"required"`
	DeliveryAvailable bool                      `json:"deliveryAvailable"`
	Images            []valueobject.ItemImage   `json:"images"`
}

// PatchItemRequest represents a request to partially update an item
type PatchItemRequest struct {
	Title             *string                    `json:"title" binding:"omitempty,min=3,max=200"`
	Description       *string                    `json:"description,omitempty"`
	PriceCents        *int                       `json:"priceCents" binding:"omitempty,gte=0"`
	Category          *string                    `json:"category,omitempty"`
	Condition         *valueobject.ItemCondition `json:"condition,omitempty"`
	Status            *valueobject.ItemStatus    `json:"status" binding:"omitempty,oneof=draft active reserved sold archived"`
	IsFeatured        *bool                      `json:"isFeatured,omitempty"`
	City              *string                    `json:"city,omitempty"`
	PostalCode        *string                    `json:"postalCode,omitempty"`
	Country           *string                    `json:"country,omitempty"`
	DeliveryAvailable *bool                      `json:"deliveryAvailable,omitempty"`
	Images            *[]valueobject.ItemImage   `json:"images,omitempty"`
}

// HasChanges returns true if at least one field is set
func (req *PatchItemRequest) HasChanges() bool {
	return req.Title != nil ||
		req.Description != nil ||
		req.PriceCents != nil ||
		req.Category != nil ||
		req.Condition != nil ||
		req.Status != nil ||
		req.IsFeatured != nil ||
		req.City != nil ||
		req.PostalCode != nil ||
		req.Country != nil ||
		req.DeliveryAvailable != nil ||
		req.Images != nil
}

// ListItemsResponse represents a paginated list of items
type ListItemsResponse struct {
	Items      []*ItemResponse `json:"items"`
	NextCursor *string         `json:"nextCursor,omitempty"`
	TotalCount int             `json:"totalCount"`
	Limit      int             `json:"limit"`
}

// ErrorResponse represents the canonical error response format
type ErrorResponse struct {
	Error *ErrorDetail `json:"error"`
}

// ErrorDetail represents error details
type ErrorDetail struct {
	Code    string                 `json:"code"`
	Message string                 `json:"message"`
	Details map[string]interface{} `json:"details,omitempty"`
}

// NewErrorResponse creates a new error response
func NewErrorResponse(code, message string, details map[string]interface{}) *ErrorResponse {
	return &ErrorResponse{
		Error: &ErrorDetail{
			Code:    code,
			Message: message,
			Details: details,
		},
	}
}

// ValidationErrorDetail represents a single validation error
type ValidationErrorDetail struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

// NewValidationErrorResponse creates a validation error response
func NewValidationErrorResponse(message string, details []ValidationErrorDetail) *ErrorResponse {
	detailsMap := make(map[string]interface{})
	for _, d := range details {
		detailsMap[d.Field] = d.Message
	}
	return NewErrorResponse("validation_error", message, detailsMap)
}

// HealthResponse represents the health check response
type HealthResponse struct {
	Status    string  `json:"status"`
	Timestamp string  `json:"timestamp"`
	Uptime    float64 `json:"uptime,omitempty"`
}

// QueryParams represents query parameters for listing items
type QueryParams struct {
	Status            *string
	Category          *string
	MinPriceCents     *int
	MaxPriceCents     *int
	City              *string
	PostalCode        *string
	IsFeatured        *bool
	DeliveryAvailable *bool
	Sort              *string
	Limit             *int
	Cursor            *string
	Query             *string // For full-text search
}

// GetLimit returns the limit with a default value
func (qp *QueryParams) GetLimit() int {
	if qp.Limit != nil && *qp.Limit > 0 {
		if *qp.Limit > 100 {
			return 100 // Max limit
		}
		return *qp.Limit
	}
	return 20 // Default limit
}

// GetCursor returns the cursor string
func (qp *QueryParams) GetCursor() string {
	if qp.Cursor != nil {
		return *qp.Cursor
	}
	return ""
}

// GetQuery returns the search query
func (qp *QueryParams) GetQuery() string {
	if qp.Query != nil {
		return *qp.Query
	}
	return ""
}
